lib/redispub/publisher.go

```diff
@@ -125,43 +136,58 @@ func PublishStream(clients []redis.UniversalClient, in <-chan *Publication, opts
 		metricsSendSuccess[i] = metricSentMessages.WithLabelValues("sent", idx)
 	}
 
+	var batchSize = config.RedisBatchSize()
+
 	for {
 		select {
 		case <-stop:
 			close(timestampC)
 			return
 
 		case p := <-in:
-			metricStalenessPreRetries.WithLabelValues(strconv.Itoa(ordinal)).Set(time.Since(p.WallTime).Seconds())
+			batch := []*Publication{p}
+			// Try to fill batch with buffer backlog
+			// (the label is needed to break out of both the select and for loop)
+		FillBatch:
+			for len(batch) < batchSize {
+				select {
+				case p2 := <-in:
+					batch = append(batch, p2)
+				default:
+					break FillBatch
+				}
+			}
+			log.Log.Debugw("Batch size", "len(batch)", len(batch))
+			metricStalenessPreRetries.WithLabelValues(strconv.Itoa(ordinal)).Set(time.Since(batch[0].WallTime).Seconds())
 			for i, publishFn := range publishFns {
-				err := publishSingleMessageWithRetries(p, 30, time.Second, publishFn)
+				err := publishBatchWithRetries(batch, 30, time.Second, publishFn)
 				log.Log.Debugw("Published to", "idx", i)
 
 				if err != nil {
-					metricsSendFailed[i].Inc()
+					metricsSendFailed[i].Add(float64(len(batch)))
 					log.Log.Errorw("Permanent error while trying to publish message; giving up",
 						"error", err,
-						"message", p)
+						"batchSize", len(batch))
 				} else {
-					metricsSendSuccess[i].Inc()
+					metricsSendSuccess[i].Add(float64(len(batch)))
 
 					// We want to make sure we do this *after* we've successfully published
 					// the messages
-					timestampC <- p.OplogTimestamp
+					timestampC <- batch[len(batch)-1].OplogTimestamp
 				}
 			}
 		}
 	}
 }
 
-func publishSingleMessageWithRetries(p *Publication, maxRetries int, sleepTime time.Duration, publishFn func(p *Publication) error) error {
-	if p == nil {
-		return errors.New("Nil Redis publication")
+func publishBatchWithRetries(batch []*Publication, maxRetries int, sleepTime time.Duration, publishFn func(batch []*Publication) error) error {
+	if len(batch) == 0 {
+		return errors.New("Empty Redis publication batch")
 	}
```

Comment:
It seems like this shouldn't be possible based on the logic of the caller? And even if the batch was empty, do we need a special error case for this? Nooping it instead seems semantically valid to me.